### PR TITLE
Changes the holodeck to use tgui, plus some code refactor/improvements

### PR DESCRIPTION
**Pull request by lzimann**
Should be slightly better, might have forgotten things because I kept doing other things while coding this.

### CODE REVIEW

**Comment by GunHog** [comment]
Make the dangerous buttons red! Does it track the current simulation? You could make that button green.

**Comment by RemieRichards** [comment]
Don't make the dangerous ones red. The Warning icon is good enough.

**Comment by lzimann** [comment]
I did try them red, did not look good/hard to read

File: code/modules/holodeck/computer.dm [hunk 1]
```diff
@@ -26,109 +29,107 @@
...
+	switch(action)
+		if("load_program")
+			var/program_to_load = text2path(params["type"])
+			if(!ispath(program_to_load))
+				return FALSE
+			var/area/A = locate(program_to_load) in GLOB.sortedAreas
+			if(A)
+				load_program(A)
+		if("safety")
+			emagged = !emagged
```
Comment (Cyberboss) [inline code/modules/holodeck/computer.dm]: Silicons can't disable emagged holodecks though?

File: code/modules/holodeck/computer.dm [hunk 2]
```diff
@@ -26,109 +29,107 @@
...
-					var/datum/effect_system/spark_spread/s = new
-					s.set_up(2, 1, T)
-					s.start()
-					return
+	if(damaged && prob(10))
+		for(var/turf/T in linked)
+			if(prob(5))
+				var/datum/effect_system/spark_spread/s = new
+				s.set_up(2, 1, T)
+				s.start()
```
Comment (Cyberboss) [inline code/modules/holodeck/computer.dm]: Use do_sparks

File: code/modules/holodeck/computer.dm [hunk 3]
```diff
@@ -144,67 +145,24 @@
...
-
-/obj/machinery/computer/holodeck/Destroy()
-	if(linked)
-		emergency_shutdown()
-		linked.linked = null
-	return ..()
+	if(!emag_programs.len)
+		to_chat(user, "[src] does not seem to have a card swipe port. It must be an inferior model.")
+		return
+	playsound(loc, 'sound/effects/sparks4.ogg', 75, 1)
```
Comment (Cyberboss) [inline code/modules/holodeck/computer.dm]: playsound(src, ...

File: code/modules/holodeck/computer.dm [hunk 4]
```diff
@@ -144,67 +145,24 @@
...
-		log_game("[key_name(user)] emagged the Holodeck Control Console")
-		updateUsrDialog()
-		nerf(!emagged)
-
-/obj/machinery/computer/holodeck/Destroy()
-	if(linked)
-		emergency_shutdown()
-		linked.linked = null
-	return ..()
+	if(!emag_programs.len)
```
Comment (Cyberboss) [inline code/modules/holodeck/computer.dm]: LAZYLEN

File: code/modules/holodeck/computer.dm [hunk 5]
```diff
@@ -216,4 +174,122 @@
...
+
+/obj/machinery/computer/holodeck/proc/nerf(active)
+	for(var/obj/item/I in spawned)
+		I.damtype = active ? STAMINA : initial(I.damtype)
+	for(var/e in effects)
+		var/obj/effect/holodeck_effect/HE = e
+		HE.safety(active)
+
+/obj/machinery/computer/holodeck/proc/load_program(area/A, force = FALSE, add_delay = TRUE)
+	if(stat)
```
Comment (Cyberboss) [inline code/modules/holodeck/computer.dm]: Not entirely sure what this check is, but if you're checking to see if it's working, use is_operational()

File: code/modules/holodeck/computer.dm [hunk 6]
```diff
@@ -216,4 +174,122 @@
...
+		return
+	if(current_cd > world.time && !force)
+		say("ERROR. Recalibrating projection apparatus.")
+		return
+	if(add_delay)
+		current_cd = world.time + HOLODECK_CD
+		if(damaged)
+			current_cd += HOLODECK_DMG_CD
+	active = (A != offline_program)
+	use_power = active + 1
```
Comment (Cyberboss) [inline code/modules/holodeck/computer.dm]: Wait for @optimumtact's PR and fix this

File: code/modules/holodeck/computer.dm [hunk 7]
```diff
@@ -216,4 +174,122 @@
...
+	if(!O)
+		return
+	var/turf/T = get_turf(O)
+	for(var/atom/movable/AM in O.contents) // these should be derezed if they were generated
+		AM.loc = T
+		if(ismob(AM))
+			silent = FALSE					// otherwise make sure they are dropped
+
+	if(!silent)
+		visible_message("The [O.name] fades away!")
```
Comment (Cyberboss) [inline code/modules/holodeck/computer.dm]: O.visible_message("[O] fades away!")

File: code/modules/holodeck/computer.dm [hunk 8]
```diff
@@ -216,4 +174,122 @@
...
+	if(O && emagged && !stat && !forced)
+		if((ismob(O) || ismob(O.loc)) && prob(50))
+			addtimer(CALLBACK(src, .proc/derez, O, silent), 50) // may last a disturbingly long time
+			return
+
+	spawned -= O
+	if(!O)
+		return
+	var/turf/T = get_turf(O)
+	for(var/atom/movable/AM in O.contents) // these should be derezed if they were generated
```
Comment (Cyberboss) [inline code/modules/holodeck/computer.dm]: in O)

File: code/modules/holodeck/computer.dm [hunk 9]
```diff
@@ -216,4 +174,122 @@
...
+		AM.loc = T
```
Comment (Cyberboss) [inline code/modules/holodeck/computer.dm]: forceMove

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/holodeck/computer.dm (hunk 1 above)
Comment (lzimann) [inline code/modules/holodeck/computer.dm]: Pretty sure they can, I can change though.

File: code/modules/holodeck/computer.dm (hunk 5 above)
Comment (lzimann) [inline code/modules/holodeck/computer.dm]: Yea, this is a leftover from the old code, didn't change this proc.

File: code/modules/holodeck/computer.dm (hunk 1 above)
Comment (Cyberboss) [inline code/modules/holodeck/computer.dm]: Probably should, seems like an oversight

Comment (GunHog) [inline code/modules/holodeck/computer.dm]: Yes, they can. There is a "Disengage" button on emagged holodecks to let them reset the safety. They can toggle it at will.

Comment (lzimann) [inline code/modules/holodeck/computer.dm]: I'll keep this behavior, better to make a new pr over this imo.

**Comment by lzimann** [comment]
Gotta wait on #28941 now

**Comment by JJRcop** [comment]
lzimann/tgstation#5

File: tgui/assets/tgui.js [hunk 10]
```diff
@@ -1,16 +1,16 @@
...",{imul:function(t,e){var n=65535,a=+t,r=+e,i=n&a,o=n&r;return 0|i*o+((n&a>>>16)*o+i*(n&r>>>16)<<16>>>0)}})},{22:22,24:24}],107:[function(t,e,n){var a=t(22);a(a.S,"Math",{log10:function(t){return Math.log(t)/Math.LN10}})},{22:22}],108:[function(t,e,n){var a=t(22);a(a.S,"Math",{log1p:t(50)})},{22:22,50:50}],109:[function(t,e,n){var a=t(22);a(a.S,"Math",{log2:function(t){return Math.log(t)/Math.LN2}})},{22:22}],110:[function(t,e,n){var a=t(22);a(a.S,"Math",{sign:t(51)})},{22:22,51:51}],111:[function(t,e,n){var a=t(22),r=t(49),i=Math.exp;a(a.S+a.F*t(24)(function(){return-2e-17!=!Math.sinh(-2e-17)}),"Math",{sinh:function(t){return Math.abs(t=+t)<1?(r(t)-r(-t))/2:(i(t-1)-i(-t-1))*(Math.E/2)}})},{22:22,24:24,49:49}],112:[function(t,e,n){var a=t(22),r=t(49),i=Math.exp;a(a.S,"Math",{tanh:function(t){var e=r(t=+t),n=r(-t);return e==1/0?1:n==1/0?-1:(e-n)/(i(t)+i(-t))}})},{22:22,49:49}],113:[function(t,e,n){var a=t(22);a(a.S,"Math",{trunc:function(t){return(t>0?Math.floor:Math.ceil)(t)}})},{22:22}],114:[function(t,e,n){"use strict";var a=t(46),r=t(29),i=t(30),o=t(11),s=t(81),u=t(24),p=t(74).trim,c="Number",l=r[c],d=l,f=l.prototype,h=o(a.create(f))==c,m="trim"in String.prototype,v=function(t){
```
Comment (Cyberboss) [inline tgui/assets/tgui.js]: Recompile

**Review by Cyberboss — DISMISSED** [review]

**Comment by Cyberboss** [comment]
@lzimann do the needful

**Comment by lzimann** [comment]
Done, tgui is also compiled.

**Pull request merged**